diff --git a/android/src/fabric/java/com/reactnativepagerview/PagerViewViewManager.kt b/android/src/fabric/java/com/reactnativepagerview/PagerViewViewManager.kt
index aad38b2bf38cec3bc9fd90301b49b368c483f673..a1de30f7aef174ec61bbd2c5657b318eb51d7286 100644
--- a/android/src/fabric/java/com/reactnativepagerview/PagerViewViewManager.kt
+++ b/android/src/fabric/java/com/reactnativepagerview/PagerViewViewManager.kt
@@ -5,6 +5,7 @@ import android.view.ViewGroup
 import androidx.viewpager2.widget.ViewPager2
 import androidx.viewpager2.widget.ViewPager2.OnPageChangeCallback
 import com.facebook.infer.annotation.Assertions
+import com.facebook.react.bridge.Arguments
 import com.facebook.react.bridge.ReactContext
 import com.facebook.react.bridge.ReadableArray
 import com.facebook.react.common.MapBuilder
@@ -38,7 +39,7 @@ class PagerViewViewManager : ViewGroupManager<NestedScrollableHost>(), RNCViewPa
     }
 
     override fun receiveCommand(root: NestedScrollableHost, commandId: String?, args: ReadableArray?) {
-        mDelegate.receiveCommand(root, commandId, args)
+        mDelegate.receiveCommand(root, commandId ?: "", args ?: Arguments.createArray())
     }
 
     public override fun createViewInstance(reactContext: ThemedReactContext): NestedScrollableHost {
